presto-orc/src/main/java/com/facebook/presto/orc/writer/SegmentedSliceBlockBuilder.java

```diff
@@ -0,0 +1,411 @@
+ */
+package com.facebook.presto.orc.writer;
+
+import com.facebook.presto.common.block.AbstractVariableWidthBlock;
+import com.facebook.presto.common.block.Block;
+import com.facebook.presto.common.block.BlockBuilder;
+import com.facebook.presto.common.block.BlockBuilderStatus;
+import com.google.common.annotations.VisibleForTesting;
+import io.airlift.slice.DynamicSliceOutput;
+import io.airlift.slice.Slice;
+import io.airlift.slice.SliceInput;
+import io.airlift.slice.XxHash64;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+import java.util.function.BiConsumer;
+
+import static io.airlift.slice.SizeOf.sizeOf;
+import static java.lang.String.format;
+
+/**
+ * Custom Block Builder implementation for use with SliceDictionaryBuilder.
+ * Instead of using one large contiguous Slice for storing the unique Strings
+ * in String dictionary, this class uses Segmented Slices. The main advantage
+ * of this class over VariableWidthBlockBuilder is memory. Non contiguous
+ * memory is more likely to be available and hence reduce the chance of OOMs.
+ * <p>
+ * Why implement a block builder ?
+ * SliceDictionaryBuilder takes in a Block and Position to write.
+ * 1. It can create a slice for the position and write it. This does not
+ * require a block builder. But temporary slice, produces lot of
+ * short lived garbage.
+ * 2. A block and position can be copied to BlockBuilder using the method
+ * Block.writeBytesTo . But this requires implementing the BlockBuilder interface.
+ * Most of the methods are going to be unused and left as Unsupported.
+ * <p>
+ * What's the difference between this class and VariableWidthBlockBuilder?
+ * This class is different from VariableWidthBlockBuilder in the following ways
+ * 1. It does not support nulls. (So null byte array and management is avoided).
+ * 2. Instead of using one contiguous chunk for storing all the entries,
+ * they are segmented.
+ * <p>
+ * How is it implemented ?
+ * The Strings from 0 to SEGMENT_SIZE-1 are stored in the first segment.
+ * The string from SEGMENT_SIZE to 2 * SEGMENT_SIZE -1 goes to the second.
+ * Each segment has Slice(data is concatenated and stored in one slice)
+ * and offsets to capture the start offset and length. New slices are appended
+ * to the open segment. Once the segment is full the segment is
+ * finalized and appended to the closed segments. A new open segment is
+ * created for further appends.
+ */
+public class SegmentedSliceBlockBuilder
+        extends AbstractVariableWidthBlock
+        implements BlockBuilder
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(SegmentedSliceBlockBuilder.class).instanceSize();
+
+    private final DynamicSliceOutput openSliceOutput;
+
+    private int openSegmentOffset;
+    private int[][] offsets;
+    private Slice[] closedSlices;
+    private long closedSlicesRetainedSize;
+    private long closedSlicesSizeInBytes;
+
+    @VisibleForTesting
+    int openSegmentIndex;
```

Comment:
Let's make it private and create a getter for it. On the getter, annotate it with `VisibleForTesting`

Reply:
Great catch, I unintentionally broke the abstraction.